src/DayPicker.tsx

```diff
@@ -256,6 +256,186 @@ export function DayPicker(props: DayPickerProps) {
+    // update previous months ref for next effect trigger
+    previousMonthsRef.current = months;
+
+    if (
+      !props.animate ||
+      !rootElementRef.current ||
+      // safety check because the ref can be set to anything by consumers
+      !(rootElementRef.current instanceof HTMLElement)
+    ) {
+      return;
+    }
+
+    // get previous root element snapshot before updating the snapshot ref
+    const previousRootElementSnapshot = previousRootElementSnapshotRef.current;
+
+    // update snapshot for next effect trigger
+    const rootElementSnapshot = rootElementRef.current.cloneNode(true);
+    if (rootElementSnapshot instanceof HTMLElement) {
+      previousRootElementSnapshotRef.current = rootElementSnapshot;
+    } else {
+      previousRootElementSnapshotRef.current = null;
+    }
+
+    // validation required for the animation to work as expected
+    if (
+      months.length === 0 ||
+      previousMonths.length === 0 ||
+      months.length !== previousMonths.length ||
+      // skip animation if a day is focused because it can cause issues to the animation and is better for a11y
+      focused
+    ) {
+      return;
+    }
+
+    const isSameMonth = dateLib.isSameMonth(
+      months[0].date,
+      previousMonths[0].date
+    );
+
+    if (isSameMonth) {
+      return;
+    }
+
+    const previousMonthElements = [
+      ...(previousRootElementSnapshot?.querySelectorAll(
+        `[data-month-container]`
+      ) ?? [])
+    ];
+
+    const currentMonthElements = [
+      ...(rootElementRef.current.querySelectorAll(`[data-month-container]`) ??
+        [])
+    ];
+
+    if (
+      currentMonthElements &&
+      currentMonthElements.every((element) => element instanceof HTMLElement) &&
+      previousMonthElements &&
+      previousMonthElements.every((element) => element instanceof HTMLElement)
+    ) {
+      const cleanUpFunctions: (() => void)[] = [];
+
+      const isAfterPreviousMonth = dateLib.isAfter(
+        months[0].date,
+        previousMonths[0].date
+      );
+
+      currentMonthElements.forEach((currentMonthElement, index) => {
+        const previousMonthElement = previousMonthElements[index];
+
+        if (!previousMonthElement) {
+          return;
+        }
+
+        // animate new displayed month
+        const monthCaptionAnimationClass = isAfterPreviousMonth
+          ? classNames[AnimationClass.animation_new_month_caption_is_after]
+          : classNames[AnimationClass.animation_new_month_caption_is_before];
+
+        const monthAnimationClass = isAfterPreviousMonth
+          ? classNames[AnimationClass.animation_new_month_is_after]
+          : classNames[AnimationClass.animation_new_month_is_before];
+
+        currentMonthElement.style.position = "relative";
+        currentMonthElement.style.overflow = "hidden";
+        const monthCaptionElement = currentMonthElement.querySelector(
+          `[data-month-caption-container]`
+        );
```

Comment:
Suggesting some refactoring for readability (maybe after we move the code to its own Hook and test it):

- `monthCaption` → just use `caption` in any instance (reasoning: there are no other "captions" in DayPicker, so the "month" prefix is superfluous)
- once we’ve moved the code to another hook, it’s evident that it’s intended for animation, so we can omit this "descriptor".
- `*Element` → `*El`

```suggestion
        const captionClassName = isAfterPreviousMonth
          ? classNames[Animation.animation_new_month_caption_is_after]
          : classNames[Animation.animation_new_month_caption_is_before];

        const monthClassName = isAfterPreviousMonth
          ? classNames[Animation.animation_new_month_is_after]
          : classNames[Animation.animation_new_month_is_before];

        currentMonthEl.style.position = "relative";
        currentMonthEl.style.overflow = "hidden";

        const captionEl = currentMonthEl.querySelector(`[data-caption-container]`);
```

Reply:
Thnks for the feedback 👍 
One last point, I just chose `monthCaption` to follow the component name `MonthCaption`, I think it's easier to associate with it by following the exact name.
I'll change in any case, let me know if you change your mind.